src/extension/type_scheme.rs

```diff
@@ -0,0 +1,269 @@
+    /// be used to create operations; at this point we only need the types.
+    ///
+    /// #Errors
+    /// Validates that all types in the schema are well-formed and all variables in the body
+    /// are declared with [TypeParam]s that guarantee they will fit.
+    pub fn new(
+        params: impl Into<Vec<TypeParam>>,
+        body: FunctionType,
+        extension_registry: &ExtensionRegistry,
+    ) -> Result<Self, SignatureError> {
+        let params = params.into();
+        body.validate(extension_registry, &params)?;
+        Ok(Self { params, body })
+    }
+
+    pub(super) fn compute_signature(
+        &self,
+        args: &[TypeArg],
+        extension_registry: &ExtensionRegistry,
+    ) -> Result<FunctionType, SignatureError> {
+        check_type_args(args, &self.params)?;
+        Ok(self.body.substitute(extension_registry, args))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::num::NonZeroU64;
+
+    use smol_str::SmolStr;
+
+    use crate::extension::prelude::{PRELUDE_ID, USIZE_T};
+    use crate::extension::{
+        ExtensionId, ExtensionRegistry, SignatureError, TypeDefBound, TypeParametrised, PRELUDE,
+        PRELUDE_REGISTRY,
+    };
+    use crate::std_extensions::collections::{EXTENSION, LIST_TYPENAME};
+    use crate::types::type_param::{TypeArg, TypeArgError, TypeParam};
+    use crate::types::{CustomType, FunctionType, Type, TypeBound};
+    use crate::Extension;
+
+    use super::OpDefTypeScheme;
+
+    #[test]
+    fn test_opaque() -> Result<(), SignatureError> {
+        let list_def = EXTENSION.get_type(&LIST_TYPENAME).unwrap();
+        let tyvar = TypeArg::use_var(0, TypeParam::Type(TypeBound::Any));
+        let list_of_var = Type::new_extension(list_def.instantiate_concrete([tyvar.clone()])?);
+        let reg: ExtensionRegistry = [PRELUDE.to_owned(), EXTENSION.to_owned()].into();
+        let list_len = OpDefTypeScheme::new(
+            [TypeParam::Type(TypeBound::Any)],
+            FunctionType::new(vec![list_of_var], vec![USIZE_T]),
+            &reg,
+        )?;
+
+        let t = list_len.compute_signature(&[TypeArg::Type { ty: USIZE_T }], &reg)?;
+        assert_eq!(
+            t,
+            FunctionType::new(
+                vec![Type::new_extension(
+                    list_def
+                        .instantiate_concrete([TypeArg::Type { ty: USIZE_T }])
+                        .unwrap()
+                )],
+                vec![USIZE_T]
+            )
+        );
+
+        Ok(())
+    }
+
+    fn id_fn(t: Type) -> FunctionType {
+        FunctionType::new(vec![t.clone()], vec![t])
+    }
+
+    #[test]
+    fn test_mismatched_args() -> Result<(), SignatureError> {
+        let ar_def = PRELUDE.get_type("array").unwrap();
+        let typarams = [TypeParam::Type(TypeBound::Any), TypeParam::max_nat()];
+        let tyvar = TypeArg::use_var(0, typarams[0].clone());
+        let szvar = TypeArg::use_var(1, typarams[1].clone());
```

Comment:
Is this what they call Polish Notation? 🤪

Reply:
Yes ;). Just for that, I've combined the redundant code in the two `let`s ;-)